### ISSUE TITLE
Heap out of memory error on large number of queued jobs

### ISSUE THREAD

**Issue by ajwootto** [OptimalBits/bull#1110]
Hi,

I currently have around 2 million tasks queued in Bull which all need to be processed. I'm trying to test the performance of a single worker against this queue of tasks, but shortly after I start the worker I get the following heap memory error:

```
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory

<--- Last few GCs --->

[13:0x2d0a5c0]    82667 ms: Mark-sweep 947.4 (1508.8) -> 947.3 (1510.8) MB, 877.4 / 0.0 ms  allocation failure GC in old space requested
[13:0x2d0a5c0]    83565 ms: Mark-sweep 947.3 (1510.8) -> 947.2 (1456.8) MB, 897.2 / 0.0 ms  last resort GC in old space requested
[13:0x2d0a5c0]    84474 ms: Mark-sweep 947.2 (1456.8) -> 947.2 (1436.3) MB, 909.3 / 0.0 ms  last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x142c70c25879 <JSObject>
    1: _settlePromise [/src/node_modules/bluebird/js/release/promise.js:~542] [pc=0xac62245a421](this=0x38989dfe5681 <Promise map = 0x3f8459124389>,promise=0x38989dfed891 <Promise map = 0x3f8459124389>,handler=0x1756049822d1 <undefined>,receiver=0x1756049822d1 <undefined>,value=0x173509168221 <Number 1.5411e+12>)
    2: _drainQueue(aka _drainQueue) [/src/node_modules/bluebird/js/release/async.js...

 1: node::Abort() [node /src/dist/app.js]
 2: 0x8cbf4c [node /src/dist/app.js]
 3: v8::Utils::ReportOOMFailure(char const*, bool) [node /src/dist/app.js]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [node /src/dist/app.js]
 5: v8::internal::Factory::NewUninitializedFixedArray(int) [node /src/dist/app.js]
 6: 0xd801bc [node /src/dist/app.js]
 7: 0xd97a95 [node /src/dist/app.js]
 8: v8::internal::JSObject::AddDataElement(v8::internal::Handle<v8::internal::JSObject>, unsigned int, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::internal::Object::ShouldThrow) [node /src/dist/app.js]
 9: v8::internal::Object::AddDataProperty(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::internal::Object::ShouldThrow, v8::internal::Object::StoreFromKeyed) [node /src/dist/app.js]
10: v8::internal::Object::SetProperty(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::LanguageMode, v8::internal::Object::StoreFromKeyed) [node /src/dist/app.js]
11: v8::internal::Runtime_SetProperty(int, v8::internal::Object**, v8::internal::Isolate*) [node /src/dist/app.js]
12: 0xac6223042fd
```

To try and narrow down the problem I've simplified my code down to the point where the process function isn't actually doing anything, it's just returning immediately. This leads me to believe there's an issue in the internals of the library when this many tasks are queued at once. I have also tried artificially limiting the speed of the processor using a timeout promise that resolves after 100ms, but I still hit this issue after 2-3 jobs are processed.

**Comment by manast** [comment]
The amount of elements in the queue should not affect the memory consumption, the queue is stored in redis, and so not affected by its memory consumption unless they are running in the same machine.

**Comment by manast** [comment]
> This leads me to believe there's an issue in the internals of the library when this many tasks are queued at once.

How does the loop that puts the jobs in the queue look like?

**Comment by ajwootto** [comment]
The jobs are put in the queue in response to an incoming HTTP request. The code looks like this:
```
    await queue.add(data, {
      delay: scheduler.delayForDate(data.date),
      attempts: 3,
      removeOnComplete: true
    })
```
However, I have it set up so there's a "broker" enqueueing jobs and a "worker", which is a separate Node process, that is actually processing the jobs. It's the worker which is having memory issues, the broker is working fine.

I'm starting parallel processors in the worker app like this:
```javascript
  queue.process(10, __dirname + '/server-src/processor.js')
```

And the process function itself is pretty simple:
```javascript
module.exports = async function(job) {
  const { data } = job
  const { date, type, eventsInfo, project_id, project_token, deliveryInfo, timezone, lastTimezone } = data
  const deliveryDate = moment(date).toDate()
  Log.debug('Processing job!')

  async function process() {
    if (type === SCHEDULE_TYPES.EVENT) {
      await logEvent({ eventsInfo, date: deliveryDate, project_id, project_token })
    } else if (type === SCHEDULE_TYPES.DELIVERY) {
      await deliverMessage({ project_id, project_token, date, deliveryInfo, timezone, lastTimezone })
    } else if (type === SCHEDULE_TYPES.SCHEDULE) {
      await triggerSchedule({ project_id, project_token, date, deliveryInfo, timezone, lastTimezone })
    } else {
      Log.warning(`Unrecognized schedule type ${type}, deleting message forever.`)
    }
  }

  await Metrics.profile(process(), 'process_schedule_job')
  processedCount++
}
```

The "logEvent" method is what is getting called from the above:

```javascript
exports.logEvent = async function({ eventsInfo, project_token, date }) {
  if (!eventsInfo) {
    return Log.warning('Missing event info to log event')
  }

  const { events, appUsers } = eventsInfo
  if (!events.length) {
    return Log.warning('Scheduled task contains no events to deliver!')
  }

  events.forEach((event) => {
    event.date = moment(date)
      .toDate()
      .getTime()
  })

  const eventPayload = Events.formatEventsBatch({ events, appUsers }, project_token)

  Log.debug(() => [`Saving ${events.length} event batches`])

  await queue.publishToQueue(eventPayload, ACTION_QUEUES[SCHEDULE_TYPES.EVENT])
  Log.debug(() => [`Saved ${events.length} event batches`])
}
```

I've also tried with all this processing code removed, and having the processor function just be empty, and the issue remains the same.

**Comment by ajwootto** [comment]
As an added bit of info, I just tried swapping Bull out with [Bee](https://github.com/bee-queue/bee-queue), and ran the same tests under the same load. Bee did not exhibit this memory usage behaviour and the worker is happily chugging through the waiting jobs.

**Comment by manast** [comment]
@ajwootto can you write a code that reproduces the issue? you mentioned above that you get the problem even with an empty processor, so why are you pasting much more complex processor code in your examples above?

**Comment by artem-itdim** [comment]
## Description
Memory  leak detected, run worker & producer in separate terminals... Open htop  and you seeing a huge list of master.js processes.

## Minimal, Working Test code to reproduce the issue.
#### (An easy to reproduce test case will dramatically decrease the resolution time.)
#### producer.js
```javascript
'use strict'

const STEP_DELAY = 500
const REDIS_HOST = '127.0.0.1'
const REDIS_PORT = 6379
const QUEUE_1 = 'first-queue'

const { random } = require('lodash')
const { of } = require('rxjs')
const { delay, take, repeat } = require('rxjs/operators')
const Queue = require('bull')

const createQueue = (name, opts = {}) => {
  const queue = new Queue(name, {
    redis: opts.redis || { host: REDIS_HOST, port: REDIS_PORT }
  })

  return queue
}

const firstQ = createQueue(QUEUE_1)

const tasks$ = of(null)
  .pipe(
    delay(STEP_DELAY),
    repeat()
  )
  .subscribe(() => {
    const num = random(0, 10000)

    firstQ.add('ftask-1', { num, task: 'ftask-1' }, {
      attempts: 3,
      removeOnComplete: true
    })
  })

process.on('SIGINT', () => {
  tasks$.unsubscribe()

  firstQ.close()

  process.exit(0)
})
```
#### proc1.js
```javascript
'use strict'

const REDIS_HOST = '127.0.0.1'
const REDIS_PORT = 6379
const QUEUE_2 = 'second-queue'

const Queue = require('bull')

const timeout = (ms) => new Promise(resolve => setTimeout(resolve, ms))

const createQueue = (name, opts = {}) => {
  const queue = new Queue(name, {
    redis: opts.redis || { host: REDIS_HOST, port: REDIS_PORT }
  })

  return queue
}

module.exports = async (job, done) => {
  await timeout(3000)

  const { num, task } = job.data
  const secondQ = createQueue(QUEUE_2)

  for (let i = 1; i <= num; i++) {
    secondQ.add('stask-1', { num, task: 'stask-1' }, {
      attempts: 3,
      removeOnComplete: true
    })
  }

  secondQ.close()
  done()
}
```
#### proc2.js
```javascript
'use strict'

const timeout = (ms) => new Promise(resolve => setTimeout(resolve, ms))

module.exports = async (job, done) => {
  await timeout(1000)
  done()
}
```
#### worker.js
```javascript
'use strict'

const REDIS_HOST = '127.0.0.1'
const REDIS_PORT = 6379
const QUEUE_1 = 'first-queue'
const QUEUE_2 = 'second-queue'

const { resolve } = require('path')
const Queue = require('bull')

const createQueue = (name, opts = {}) => {
  const queue = new Queue(name, {
    redis: opts.redis || { host: REDIS_HOST, port: REDIS_PORT }
  })

  return queue
}

console.log('start worker...')

const firstQ = createQueue(QUEUE_1)
const secondQ = createQueue(QUEUE_2)
const queues = [firstQ, secondQ]

firstQ.process('ftask-1', 100, resolve('proc1.js'))
secondQ.process('stask-1', 200, resolve('proc2.js'))

queues.map(queue => {
  queue
    .on('error', err => {
      console.log(`error ${queue.name}`)
      console.log(err)
    })
    .on('failed', (job, err) => {
      console.log(`failed ${queue.name}`)
      console.log(job.data)
      console.log(err)
    })
})

process.on('SIGINT', async () => {
  await Promise.all(queues.map(queue => queue.close()))

  process.exit(0)
})
```

## Bull version
3.5.1

## Additional information
nodejs: v10.13.0

```json
  "dependencies": {
    "bull": "^3.5.1",
    "lodash": "^4.17.11",
    "rxjs": "^6.3.3"
  }
```

**Comment by OkuraSt** [comment]
I have maybe 6 queues that download files and process information, each job increases memory usage but  after they finish it never gets released. I was expecting this behavior, is this correct or I need to release process memory manually?

My queues process from a file like the example above.

**Comment by artem-itdim** [comment]
Hi, can't you still reproduce it?

One more clarification: this happens only with separate processes, if we use process function all works fine.

for normal works
```javascript
firstQ.process('ftask-1', 100, require(resolve('proc1.js')))
secondQ.process('stask-1', 200, require(resolve('proc2.js')))
```

**Comment by manast** [comment]
btw, do not create a queue instance in the processor, that will also consume a lot of extra resources, like here:
```module.exports = async (job, done) => {
  await timeout(3000)

  const { num, task } = job.data
  const secondQ = createQueue(QUEUE_2)

  for (let i = 1; i <= num; i++) {
    secondQ.add('stask-1', { num, task: 'stask-1' }, {
      attempts: 3,
      removeOnComplete: true
    })
  }

  secondQ.close()
  done()
}
```

you should not use done either, these are async methods already:
```
  const secondQ = createQueue(QUEUE_2)

module.exports = async (job) => {
  await timeout(3000)

  const { num, task } = job.data

  for (let i = 1; i <= num; i++) {
    secondQ.add('stask-1', { num, task: 'stask-1' }, {
      attempts: 3,
      removeOnComplete: true
    })
  }

  secondQ.close()
}
```

**Comment by ajwootto** [comment]
Hi again, sorry I haven't had time to produce a test case, though it seems @artem-itdim did it for me

I have managed to capture heapdumps that seem to show the memory usage increase in action. Right now I can't share them directly because they have sensitive data in them, but the major culprit seems to be string allocations for eval commands as shown in this screenshot:

![image](https://user-images.githubusercontent.com/1740731/49248165-ea53b600-f3cd-11e8-9fdd-8eb7a9ee6c6c.png)

Also, I believe the problem is most likely to happen when there are millions of jobs whose schedules are up and need to be moved from "delayed" to "waiting". To reproduce, enqueue a few million jobs scheduled to go out in 1 second but don't provide a processor function. Then run a processor which starts moving the jobs from delayed to waiting. During this process, I've observed the memory usage climbing without bound.

**Comment by kartojal** [comment]
Have a similar issue here, regarding memory usage.
Have 4 repeatable tasks, each one have a separate process file. Every task does one HTTP request and then saves data to a mysql database.

Each task is repeated every 15 seconds, and the memory is never released again after finishing the task. The memory keeps climbing until the server runs out of RAM, and can see multiple "master.js" proccesses spawned, first there was 4 master processes, now there is  7 processes after 1 hour running.

Any "good practices" for repeteable tasks to prevent memory usage leaks?

**Comment by rsaphala** [comment]
related issue #1008
Any updates on this?

**Comment by manast** [comment]
@rsaphala we need a reproducible test case... so far I am not able to reproduce these issue or issues.

**Comment by redhat-raptor** [comment]
Here is how the issue can be replicated:

### Producer:
```
var Queue = require('bull');
var sendQueue = new Queue("test");

(async (limit) => {
  for(let i=0; i<limit; ++i) {
     await sendQueue.add({msg: 'Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur...'}, {
         removeOnComplete: true,
         delay: 5 * (60 * 1000)
     });
  }

  console.log('Done!');
  process.exit(0);
})(5000000);

```

Once producer is done creating jobs, run the consumer:

### Consumer:
```
var Queue = require('bull');
var receiveQueue = new Queue("test");

receiveQueue.process(function(job, done){
    console.log("Received message", job.id);
    done();
});
```

this will cause the memory consumption to go crazy and eventually stop processing jobs and hang.
Tested with bull version `"bull": "^3.6.0"`

**Comment by rsaphala** [comment]
> @manast Here is how the issue can be replicated:
> 
> ### Producer:
> ```
> var Queue = require('bull');
> var sendQueue = new Queue("test");
> 
> (async (limit) => {
>   for(let i=0; i<limit; ++i) {
>      await sendQueue.add({msg: 'Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur...'}, {
>          removeOnComplete: true,
>          delay: 5 * (60 * 1000)
>      });
>   }
> 
>   console.log('Done!');
>   process.exit(0);
> })(5000000);
> ```
> Once producer is done creating jobs, run the consumer:
> 
> ### Consumer:
> ```
> var Queue = require('bull');
> var receiveQueue = new Queue("test");
> 
> receiveQueue.process(function(job, done){
>     console.log("Received message", job.id);
>     done();
> });
> ```
> this will cause the memory consumption to go crazy and eventually stops processing jobs and hangs.
> Tested with bull version `"bull": "^3.6.0"`

I can confirm that this indeed causes out of memory error. @manast Can you take a look into this? I

**Comment by redhat-raptor** [comment]
@davidsteinberger The producer is not a sandboxed process.

**Comment by kartojal** [comment]
@redhat-raptor  yes also happens to me, but the memory leak is much less slower compared to new versions, at least in my setup with Bull 3.4.8 in single process mode, PM2 3.2.2 and node 8.

**Comment by manast** [comment]
The memory leak could be related to the removal of Bluebird. Some promise that for some reason is dangling somewhere. I will take a look at it as soon as possible.

**Comment by manast** [comment]
ok. So have checked this, and it is a problem that happens with delayed jobs, specifically if you have many delayed jobs that should be processed at the same time. The current implementation is not efficient at all for this use case, which creates this problem, not sure if it really is a leak, or just the inefficient implementation. I am working on a solution that should process delayed jobs faster for this edge case. I will keep you updated.

**Comment by brian-mann** [comment]
Can confirm this causes a major bug in that if a huge number of jobs are rate limited then bull will actually **stop** picking up jobs from this queue permanently.

I spent close to 8 hours debugging and it appears as if there is a race condition that causes the `Number.MAX_VALUE` to be set in a timer somewhere which prevents the queue from ever recovering (like its supposed to in the guardian function that runs after 5 seconds).

The problem manifests itself where you have a concurrency value set to anything above 1. It appears that when the concurrency jobs spin up - they are blocked by `getNewJob` because the jobs are delayed. However, there is a race condition that emerges where bull eventually becomes locked and stops checking for new jobs permanently. I can't confirm the precise location of the code where this is happening, but it is 100% reproducible given several conditions:

- you set concurrency of the job higher than 1 - although the higher you set it the faster the bug will reveal itself when rate limiting begins
- you set a limiter with a number + duration that allows for say a dozen or two events to be processed before the rate limiter takes effect
- you have enough jobs already in the waiting state - in our case it was several hundreds of thousands. this number likely **doesn't** need to be this high, it just needs to be high enough that the rate limiter comes into play

What will happen is that after jobs are delayed, they will move back to the `waiting` queue and then never get picked up again by any workers.

In order to drain our backlog of events, I could have reduced the concurrency down to `1` (although that would have taken forever) so we opted to remove the rate limiting altogether after ensuring the 3rd party we were interacting with could handle the huge backlog of events. Removing the rate limiting fixed the problem.

**Comment by brian-mann** [comment]
@manast 	updated my comment to provide more context ^^

**Comment by manast** [comment]
I am working on the last touches to the fix in the delay logic, hopefully will solve the issue. I will also try to reproduce your case to verify that the fix really helps.

**Comment by manast** [comment]
It is late so I am not gonna have time to test it today, but if anybody has the time, here is the complete PR: https://github.com/OptimalBits/bull/pull/1207

**Comment by manast** [comment]
@brian-mann The issue with rate-limiting jobs will be alleviated with a new PR I am working on (https://github.com/OptimalBits/bull/pull/1212), however for an ultimate solution we will need to implement this: https://github.com/OptimalBits/bull/issues/1211
Could be a few weeks from now due to my current work load.

**Comment by abriosi** [comment]
#1222

**Comment by bengawel** [comment]
Any updates on when we can expect this to be implemented?

**Comment by celiolatorraca** [comment]
Is this quick-fix available in the latest version (3.7.0)?

**Comment by manast** [comment]
@celiolatorraca are you referring to rate limiter or standard delayed jobs?

**Comment by celiolatorraca** [comment]
Hi @manast !!

I thought I had the same problem as described here in this issue... But I just discovered that I'm having a High CPU usage when queueing jobs...
I've tried setting concurrency to 1 because to see if it was related to Rate Limiter... But no luck!

I've tried some profiling on my code to understand why, and I got that we are spending too many time deciding which job will be executed

Do you have any idea if it's somehow related to this? If not, should I open another issue?

Here is a part of the profile we captured:
```
[Summary]:
   ticks  total  nonlib   name
  127995   21.4%   82.9%  JavaScript
      0    0.0%    0.0%  C++
  129609   21.7%   84.0%  GC
  442575   74.1%          Shared libraries
  26384    4.4%          Unaccounted

 [C++ entry points]:
   ticks    cpp   total   name

 [Bottom up (heavy) profile]:
  Note: percentage shows a share of a particular caller in the total
  amount of its parent calls.
  Callers occupying less than 1.0% are not shown.

   ticks parent  name
  273478   45.8%  /usr/local/bin/node
  120429   44.0%    /usr/local/bin/node
  27043   22.5%      Stub: CallApiCallbackStub
  23326   86.3%        LazyCompile: *_tickCallback internal/process/next_tick.js:41:25
    637    2.4%        LazyCompile: *clearBuffer _stream_writable.js:495:21
  24656   20.5%      Builtin: PromisePrototypeThen
  14950   60.6%        LazyCompile: *scripts.updateDelaySet.then.nextTimestamp /opt/app/services/logistics-kafka-consumer/node_modules/bull/lib/queue.js:877:15
  14709   98.4%          Stub: CallApiCallbackStub
  11902   80.9%            LazyCompile: *_tickCallback internal/process/next_tick.js:41:25
    477    3.2%            LazyCompile: *clearBuffer _stream_writable.js:495:21
    190    1.3%            LazyCompile: *onStreamRead internal/stream_base_commons.js:87:22
    241    1.6%          /usr/local/bin/node
    241  100.0%            LazyCompile: ~_tickCallback internal/process/next_tick.js:41:25
   4632   18.8%        LazyCompile: *delayUpdate /opt/app/services/logistics-kafka-consumer/node_modules/bull/lib/queue.js:874:25
   3829   82.7%          LazyCompile: *update /opt/app/services/logistics-kafka-consumer/node_modules/bull/lib/queue.js:865:20
   2857   74.6%            Stub: CallApiCallbackStub
    652   17.0%            LazyCompile: *eclient.on /opt/app/services/logistics-kafka-consumer/node_modules/bull/lib/queue.js:439:30
    295    7.7%            LazyCompile: *scripts.updateDelaySet.then.nextTimestamp /opt/app/services/logistics-kafka-consumer/node_modules/bull/lib/queue.js:877:15
    622   13.4%          RegExp: <%-([?)%>|+?)$^)|<%([?)%>|$ {1}
    622  100.0%            LazyCompile: *processTimers timers.js:220:23
    175    3.8%          LazyCompile: *tryOnTimeout timers.js:292:22
    175  100.0%            LazyCompile: *processTimers timers.js:220:23
   1998    8.1%        LazyCompile: *<anonymous> /opt/app/services/logistics-kafka-consumer/node_modules/ioredis/built/commander.js:125:21
   1994   99.8%          LazyCompile: *delayUpdate /opt/app/services/logistics-kafka-consumer/node_modules/bull/lib/queue.js:874:25
   1679   84.2%            LazyCompile: *update /opt/app/services/logistics-kafka-consumer/node_modules/bull/lib/queue.js:865:20
    237   11.9%            RegExp: <%-([?)%>|+?)$^)|<%([?)%>|$ {1}
     71    3.6%            LazyCompile: *tryOnTimeout timers.js:292:22
   1009    4.1%        LazyCompile: *processTimers timers.js:220:23
    510    2.1%        Stub: CallApiCallbackStub
    450   88.2%          LazyCompile: *_tickCallback internal/process/next_tick.js:41:25
    485    2.0%        LazyCompile: *tryOnTimeout timers.js:292:22
    330    1.3%        LazyCompile: ~processTimers timers.js:220:23
```

**Comment by jerv13** [comment]
This might be related to the removal of 'BluebirdPromise'.  We had a similar issue (v3.3.12) and had to configure BluebirdPromise like so:
```
// If long stack traces are turned on and we setup a queue listener with concurrency,
// we will get a memory leak. Since long stack traces keep a reference to the previous
// promise and Bull uses an infinite promise chain, we need to disable long stack traces.
BluebirdPromise.config({
  longStackTraces: false,
});
```

Since BluebirdPromise is gone, this may be an issue.

**Comment by celiolatorraca** [comment]
Hi @jerv13,

Do you have any idea which version bluebirdPromise was removed?
It would be good to try the version with bluebirdPromise to see if It works...

**Comment by moliinyk** [comment]
I'm having the issue with Bull ATM. Nothing serious, container memory consumption is around ~2G all the time.

Current setup: 

8 workers 
8 different queues 
8 concurrency rate

These numbers have been carefully calculated based on the system load and average CPU usage of the worker container. CPU usage is in between 60-80%. Peak processing hours up to 60K jobs per hour.

I use a `sandboxed` processors and instantiate them only once during server start. Due to the nature of the services, we rely on "retry" mechanism that Bull provides, hence, why my implementation is quite coupled via queues. As soon as job is processed, message goes to another queue etc. But the way it's done I create a new instance of the `Queue` objects and add messages from within sandboxed a processor, which is bad.

I tried to use `process.on(`SIGTERM`, async () => { await queue.close(); })` to close Redis connections after process dies but it resulted in a NodeJS memory leak (node v8.x). So, right now when I do `await queue.add()` I actually never close Redis connection.

I don't know how to handle that properly but longer-term plan for my team is to decouple every processing step into its own service and use proper PubSub event-driven architecture.

Before I introduced `sandboxed` processors memory was around 600Mb but we had another problem of "stalled" jobs.

Maybe there is advice on how to optimize Bull & Redis connections, maybe there is a way to re-use same connection across multiple processors?

**Comment by manast** [comment]
It would be interesting if you can reproduce this behaviour with BullMQ  https://github.com/taskforcesh/bullmq
because the suspicion I have it is that the bug was introduced when removing bluebird, however in BullMQ we do use async/await instead, so maybe it works better.

**Comment by moliinyk** [comment]
@manast we've been running our Bull Queue in production for about a year now. We've processed over 
 20 million jobs in total. Observation so far is that since we introduced sandboxed processors RAM consumption has been quite stable, high bit stable (no memory leaks have been detected). During peak hours we produce over 10K concurrent connections to our Redis instance.

I think we've got a pretty stable setup atm which helps us to solve our interim goals. We consider moving to a pure event-driven architecture using PubSub queue approach.
